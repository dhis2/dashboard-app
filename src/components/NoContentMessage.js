import PropTypes from 'prop-types'
import React from 'react'
import classes from './styles/NoContentMessage.module.css'

const NoContentMessage = ({ text }) => (
    <div className={classes.container}>{text}</div>
)

NoContentMessage.propTypes = {
    text: PropTypes.string,
}

export default NoContentMessage
