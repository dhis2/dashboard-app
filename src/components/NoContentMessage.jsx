import { colors } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import classes from './styles/NoContentMessage.module.css'

const NoContentMessage = ({ text }) => (
    <div className={classes.container}>
        <svg
            width="48"
            height="48"
            viewBox="0 0 48 48"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
        >
            <path
                d="M6.4387 47.8477L6.82684 45.8858C7.20466 45.9605 7.59673 46 8 46H10V48H8C7.46572 48 6.94367 47.9476 6.4387 47.8477ZM38 48V46H40C40.4033 46 40.7953 45.9605 41.1732 45.8858L41.5613 47.8477C41.0563 47.9476 40.5343 48 40 48H38ZM44.4449 46.6525L43.3323 44.9906C43.9875 44.552 44.552 43.9875 44.9906 43.3323L46.6525 44.4449C46.0685 45.3173 45.3173 46.0685 44.4449 46.6525ZM48 10H46V8C46 7.59673 45.9605 7.20466 45.8858 6.82684L47.8477 6.4387C47.9476 6.94367 48 7.46572 48 8V10ZM46.6525 3.55507L44.9906 4.66771C44.552 4.01254 43.9875 3.44804 43.3323 3.00941L44.4449 1.34748C45.3173 1.93154 46.0685 2.68266 46.6525 3.55507ZM10 0H8C7.46572 0 6.94367 0.0523755 6.4387 0.152272L6.82684 2.11425C7.20466 2.0395 7.59673 2 8 2H10V0ZM3.55508 1.34748L4.66771 3.00941C4.01254 3.44804 3.44804 4.01254 3.00941 4.66771L1.34748 3.55507C1.93154 2.68266 2.68266 1.93154 3.55508 1.34748ZM0 38H2V40C2 40.4033 2.0395 40.7953 2.11425 41.1732L0.152272 41.5613C0.0523755 41.0563 0 40.5343 0 40V38ZM3.55507 46.6525L4.66771 44.9906C4.01254 44.552 3.44804 43.9875 3.00941 43.3323L1.34748 44.4449C1.93154 45.3173 2.68266 46.0685 3.55507 46.6525ZM0 34H2V30H0V34ZM0 26H2V22H0V26ZM0 18H2V14H0V18ZM0 10H2V8C2 7.59673 2.0395 7.20466 2.11425 6.82684L0.152272 6.4387C0.0523756 6.94367 0 7.46572 0 8V10ZM14 0V2H18V0H14ZM22 0V2H26V0H22ZM30 0V2H34V0H30ZM38 0V2H40C40.4033 2 40.7953 2.0395 41.1732 2.11425L41.5613 0.152272C41.0563 0.0523755 40.5343 0 40 0H38ZM48 14H46V18H48V14ZM48 22H46V26H48V22ZM48 30H46V34H48V30ZM48 38H46V40C46 40.4033 45.9605 40.7953 45.8858 41.1732L47.8477 41.5613C47.9476 41.0563 48 40.5343 48 40V38ZM34 48V46H30V48H34ZM26 48V46H22V48H26ZM18 48V46H14V48H18Z"
                fill={colors.grey500}
            />
        </svg>

        <p>{text}</p>
    </div>
)

NoContentMessage.propTypes = {
    text: PropTypes.string,
}

export default NoContentMessage
