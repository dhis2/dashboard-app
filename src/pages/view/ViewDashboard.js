import { useOnlineStatus } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { AlertStack, AlertBar } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { useState, useEffect } from 'react'
import { connect } from 'react-redux'
import { acClearEditDashboard } from '../../actions/editDashboard'
import { acSetPassiveViewRegistered } from '../../actions/passiveViewRegistered'
import { acClearPrintDashboard } from '../../actions/printDashboard'
import {
    tSetSelectedDashboardById,
    tSetSelectedDashboardByIdOffline,
} from '../../actions/selected'
import { apiPostDataStatistics } from '../../api/dataStatistics'
import DashboardContainer from '../../components/DashboardContainer'
import LoadingMask from '../../components/LoadingMask'
import Notice from '../../components/Notice'
import { setHeaderbarVisible } from '../../modules/setHeaderbarVisible'
import { useCacheableSection } from '../../modules/useCacheableSection'
import { sGetDashboardById } from '../../reducers/dashboards'
import { sGetPassiveViewRegistered } from '../../reducers/passiveViewRegistered'
import { sGetSelectedId } from '../../reducers/selected'
import DashboardsBar from './DashboardsBar/DashboardsBar'
import FilterBar from './FilterBar/FilterBar'
import ItemGrid from './ItemGrid'
import classes from './styles/ViewDashboard.module.css'
import TitleBar from './TitleBar/TitleBar'

const ViewDashboard = props => {
    const [controlbarExpanded, setControlbarExpanded] = useState(false)
    const [loadingMessage, setLoadingMessage] = useState(null)
    const [loaded, setLoaded] = useState(false)
    const [loadFailed, setLoadFailed] = useState(false)
    const { online } = useOnlineStatus()
    const { isCached, recordingState } = useCacheableSection(props.requestedId)

    useEffect(() => {
        setHeaderbarVisible(true)
        props.clearEditDashboard()
        props.clearPrintDashboard()
    }, [])

    useEffect(() => {
        setLoaded(false)

        Array.from(
            document.getElementsByClassName('dashboard-scroll-container')
        ).forEach(container => {
            container.scroll(0, 0)
        })
    }, [props.requestedId])

    useEffect(() => {
        if (!props.passiveViewRegistered && online) {
            apiPostDataStatistics('PASSIVE_DASHBOARD_VIEW', props.requestedId)
                .then(() => {
                    props.registerPassiveView()
                })
                .catch(error => console.info(error))
        }
    }, [props.passiveViewRegistered])

    useEffect(() => {
        const loadDashboard = async () => {
            const alertTimeout = setTimeout(() => {
                if (props.requestedDashboardName) {
                    setLoadingMessage(
                        i18n.t('Loading dashboard â€“ {{name}}', {
                            name: props.requestedDashboardName,
                        })
                    )
                } else {
                    setLoadingMessage(i18n.t('Loading dashboard'))
                }
            }, 500)

            try {
                setLoaded(true)
                await props.fetchDashboard(props.requestedId, props.username)

                setLoadFailed(false)
                setLoadingMessage(null)
                clearTimeout(alertTimeout)
            } catch (e) {
                setLoaded(false)
                setLoadFailed(true)
                setLoadingMessage(null)
                clearTimeout(alertTimeout)
                props.setSelectedAsOffline(props.requestedId, props.username)
            }
        }

        const requestedIsAvailable = online || isCached
        const switchingDashboard = props.requestedId !== props.currentId
        if (
            requestedIsAvailable &&
            (recordingState === 'recording' || !loaded)
        ) {
            loadDashboard()
        } else if (!requestedIsAvailable && switchingDashboard) {
            setLoaded(false)
            props.setSelectedAsOffline(props.requestedId, props.username)
        }
    }, [props.requestedId, props.currentId, loaded, recordingState, online])

    const onExpandedChanged = expanded => setControlbarExpanded(expanded)

    const getContent = () => {
        if (
            !online &&
            !isCached &&
            (props.requestedId !== props.currentId || !loaded)
        ) {
            return (
                <Notice
                    title={i18n.t('Offline')}
                    message={i18n.t(
                        'This dashboard cannot be loaded while offline.'
                    )}
                />
            )
        }

        if (loadFailed) {
            return (
                <Notice
                    title={i18n.t('Load dashboard failed')}
                    message={i18n.t(
                        'This dashboard could not be loaded. Please try again later.'
                    )}
                />
            )
        }

        return props.requestedId !== props.currentId ? (
            <LoadingMask />
        ) : (
            <>
                <TitleBar />
                <FilterBar />
                <ItemGrid />
            </>
        )
    }

    return (
        <>
            <div
                className={cx(classes.container, 'dashboard-scroll-container')}
                data-test="outer-scroll-container"
            >
                <DashboardsBar
                    expanded={controlbarExpanded}
                    onExpandedChanged={onExpandedChanged}
                />
                <DashboardContainer covered={controlbarExpanded}>
                    {controlbarExpanded && (
                        <div
                            className={classes.cover}
                            onClick={() => setControlbarExpanded(false)}
                        />
                    )}
                    {getContent()}
                </DashboardContainer>
            </div>
            <AlertStack>
                {loadingMessage && (
                    <AlertBar
                        onHidden={() => setLoadingMessage(null)}
                        permanent
                    >
                        {loadingMessage}
                    </AlertBar>
                )}
            </AlertStack>
        </>
    )
}

ViewDashboard.propTypes = {
    clearEditDashboard: PropTypes.func,
    clearPrintDashboard: PropTypes.func,
    currentId: PropTypes.string,
    fetchDashboard: PropTypes.func,
    passiveViewRegistered: PropTypes.bool,
    registerPassiveView: PropTypes.func,
    requestedDashboardName: PropTypes.string,
    requestedId: PropTypes.string,
    setSelectedAsOffline: PropTypes.func,
    username: PropTypes.string,
}

const mapStateToProps = (state, ownProps) => {
    const dashboard = sGetDashboardById(state, ownProps.requestedId) || {}

    return {
        passiveViewRegistered: sGetPassiveViewRegistered(state),
        requestedDashboardName: dashboard.displayName || null,
        currentId: sGetSelectedId(state),
    }
}

export default connect(mapStateToProps, {
    clearEditDashboard: acClearEditDashboard,
    clearPrintDashboard: acClearPrintDashboard,
    registerPassiveView: acSetPassiveViewRegistered,
    fetchDashboard: tSetSelectedDashboardById,
    setSelectedAsOffline: tSetSelectedDashboardByIdOffline,
})(ViewDashboard)
