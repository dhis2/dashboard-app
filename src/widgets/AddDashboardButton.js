import React from 'react';
import AddIcon from '@material-ui/icons/AddCircle';
import { colors } from '../modules/colors';
import { NONAME } from 'dns';

const styles = {
    button: {
        width: 36,
        height: 36,
        padding: 0,
        backgroundColor: 'transparent',
        border: 'none',
        boxShadow: 'none',
        // cursor: 'pointer',
    },
    icon: {
        fill: colors.mediumGreen,
        // color: colors.mediumGreen,
        width: '1.3em',
        height: '1.3em',
        // cursor: 'pointer',
    },
};

const AddButton = () => {
    return (
        <button style={styles.button}>
            <AddIcon style={styles.icon} />
        </button>
    );
};

export default AddButton;
