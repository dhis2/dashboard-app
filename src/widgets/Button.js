import React from 'react';
import FlatButton from 'material-ui/FlatButton';
import { colors } from '../colors';

const style = {
    color: colors.royalBlue,
    backgroundColor: 'transparent',
    border: 'none',
    fontSize: '14px',
    fontWeight: 500,
    height: '36px',
    cursor: 'pointer',
};

const PlainButton = ({ onClick, children }) => {
    const label = typeof children === 'string' ? children : null;

    return (
        <FlatButton
            style={style}
            onClick={onClick}
            label={label}
            disabledBackgroundColor={colors.white}
            disabledLabelColor={colors.lightMediumGrey}
        />
    );
};

export default PlainButton;
