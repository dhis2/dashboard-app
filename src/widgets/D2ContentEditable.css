[contenteditable='true']:empty:before {
    content: attr(data-text);
    color: #999;
    font-weight: 400;
}

.editable-text:focus {
    outline: 0 none;
    cursor: text;
}
