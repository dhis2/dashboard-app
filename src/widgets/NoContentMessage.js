import React from 'react';
import { colors } from '../colors';

export const NoContentMessage = ({ text }) => (
    <div
        style={{
            padding: '50px 10px',
            textAlign: 'center',
            fontSize: '15px',
            fontWeight: 500,
            color: colors.charcoalGrey,
        }}
    >
        {text}
    </div>
);

export default NoContentMessage;
