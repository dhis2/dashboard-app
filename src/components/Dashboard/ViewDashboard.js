import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import isEmpty from 'lodash/isEmpty';
import i18n from '@dhis2/d2-i18n';

import {
    sGetAllDashboards,
    sDashboardsIsFetching,
} from '../../reducers/dashboards';
import { sGetSelectedId } from '../../reducers/selected';
import DashboardsBar from '../ControlBar/DashboardsBar';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import DashboardContent from './DashboardContent';
import NoContentMessage from '../../widgets/NoContentMessage';

export const Content = ({ hasDashboardContent, dashboardsIsEmpty }) => {
    const msg = dashboardsIsEmpty
        ? i18n.t(
              'No dashboards found. Use the + button to create a new dashboard.'
          )
        : i18n.t('Requested dashboard not found');

    return hasDashboardContent ? (
        <DashboardContent editMode={false} />
    ) : (
        <NoContentMessage text={msg} />
    );
};

export const ViewDashboard = ({ id, dashboardsIsEmpty, dashboardsLoaded }) => {
    const hasDashboardContent = id && !dashboardsIsEmpty;
    const contentNotReady = !dashboardsLoaded || id === null;

    return (
        <Fragment>
            <DashboardsBar />
            <DashboardVerticalOffset />
            <div className="dashboard-wrapper">
                {contentNotReady ? null : (
                    <Content
                        hasDashboardContent={hasDashboardContent}
                        dashboardsIsEmpty={dashboardsIsEmpty}
                    />
                )}
            </div>
        </Fragment>
    );
};

const mapStateToProps = state => {
    const dashboards = sGetAllDashboards(state);

    return {
        id: sGetSelectedId(state),
        dashboardsIsEmpty: isEmpty(dashboards),
        dashboardsLoaded: !sDashboardsIsFetching(state),
    };
};

export default connect(mapStateToProps)(ViewDashboard);
