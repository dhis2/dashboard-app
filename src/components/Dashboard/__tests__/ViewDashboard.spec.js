import React from 'react';
import { shallow } from 'enzyme';
import { ViewDashboard, Content } from '../ViewDashboard';
import { NoContentMessage } from '../../../widgets/NoContentMessage';

jest.mock('../DashboardContent', () => () => <div />);
jest.mock('../../ControlBar/DashboardsBar', () => () => <div />);
jest.mock('../DashboardVerticalOffset', () => () => <div />);

describe('ViewDashboard', () => {
    let props;
    let shallowViewDashboard;
    const viewDashboard = () => {
        if (!shallowViewDashboard) {
            shallowViewDashboard = shallow(<ViewDashboard {...props} />);
        }
        return shallowViewDashboard;
    };

    const assertContent = hasContent => {
        const content = viewDashboard().find(Content);

        expect(content.length).toBe(1);
        expect(content.dive().find(NoContentMessage)).toHaveLength(
            hasContent ? 0 : 1
        );
    };

    beforeEach(() => {
        props = {
            id: undefined,
            dashboardsIsEmpty: undefined,
            dashboardsLoaded: undefined,
        };
        shallowViewDashboard = undefined;
    });

    describe('when "dashboardsLoaded" is false', () => {
        it('does not render any children inside the div', () => {
            props.dashboardsLoaded = false;

            expect(
                viewDashboard()
                    .find('.dashboard-wrapper')
                    .children().length
            ).toBe(0);
        });
    });

    describe('when "dashboardsLoaded" is true', () => {
        beforeEach(() => {
            props.dashboardsLoaded = true;
        });

        describe('when "id" is null', () => {
            it('does not render any children inside the div', () => {
                props.id = null;
                expect(
                    viewDashboard()
                        .find('.dashboard-wrapper')
                        .children().length
                ).toBe(0);
            });
        });

        describe('when "dashboardsIsEmpty" is true', () => {
            beforeEach(() => {
                props.dashboardsIsEmpty = true;
            });

            it('renders a NoContentMessage', () => {
                assertContent(false);
            });
        });

        describe('when "dashboardsIsEmpty" is false', () => {
            beforeEach(() => {
                props.dashboardsIsEmpty = false;
            });

            describe('when id is not null or false', () => {
                beforeEach(() => {
                    props.id = '123xyz';
                });

                it('renders DashboardContent', () => {
                    assertContent(true);
                });
            });

            describe('when id is false', () => {
                beforeEach(() => {
                    props.id = false;
                });

                it('renders a NoContentMessage', () => {
                    assertContent(false);
                });
            });
        });
    });
});
