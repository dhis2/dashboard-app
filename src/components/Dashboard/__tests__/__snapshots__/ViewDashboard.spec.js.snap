// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewDashboard renders correctly default 1`] = `
<Fragment>
  <DashboardsBar />
  <div
    className="dashboard-wrapper"
    style={
      Object {
        "height": 448,
      }
    }
  >
    <TitleBar
      edit={false}
    />
    <FilterBar />
    <ItemGrid
      edit={false}
    />
  </div>
</Fragment>
`;
